Feat:Code Organization (#25)
fsprojects/fsharp-cheatsheet: docs/fsharp-cheatsheet.md

    let square x = x * x
    let print x = printfn "The number is: %d" x

>     let squareNegateThenPrint x =

Reviewer: needs to mvoe to L129

Author: Sorry, I'm not understanding this comment. Here is lines 127 to 131:

```
127 Pipe operator `|>` is used to chain functions and arguments together. Double-backtick identifiers are handy to improve readability especially in unit testing:
128
129    let ``square, negate, then print`` x =
130       x |> square |> negate |> print
131
```

Reviewer: was saying that there is/was `print (square (negate x))`, then `print x |> square |> negate`, then `print >> square >> negate`
And one of the fns has `'` (prime) naming as it was the second.

if you remove the very first one, then the motivaiton for the second is lost
so prob best to move it down and work it into that section

(and consider the flow onward to showing it with `>>`)

This operator can assist the F# type checker by providing type information before use:

    let sumOfLengths (xs : string []) =
        xs
        |> Array.map (fun s -> s.Length)
        |> Array.sum

Composition operator `>>` is used to compose functions:

>     let squareNegateThenPrint' =

Reviewer: this was a ref to L125 a) illustrating `'` b) illustrating the correspondence between `>>` and parenthesized nested calls

Author: I cleared all trailing white space. Didn't realize how man lines were changed until after. :-/

*Partial active patterns*:

    let (|DivisibleBy|_|) by n =
        if n % by = 0 then Some DivisibleBy else None

    let fizzBuzz = function
        | DivisibleBy 3 & DivisibleBy 5 -> "FizzBuzz"
        | DivisibleBy 3 -> "Fizz"
        | DivisibleBy 5 -> "Buzz"
        | i -> string i

*Partial active patterns* share the syntax of parameterized patterns but their active recognizers accept only one argument.

<a name="CodeOrganization"></a>Code Organization
---------

### Modules

> Modules are key building blocks for grouping related concepts; they can contain `types`, `let` bindings, and even other `modules`.

Reviewer: ```suggestion
Modules are key building blocks for grouping related code; they can contain `types`, `let` bindings, and container (nest) other `module`s.
```

> Modules can be referenced using dot notation or exposed with the `open` keyword. Illustrative-only example:

Reviewer: ```suggestion
Identifiers within modules can be referenced using dot notation, or you can bring them into scope via the `open` keyword. Illustrative-only example:
```

    module Game =
        let mutable basePoints = 1
        type Player = { id: int; score: int }
        let playerScored player = { player with score = player.score + basePoints }

    let player: Game.Player = { id = 1; score = 0 }
    let player' = Game.playerScored player  // score = 1

    open Game
    basePoints <- 2
    let player'' = playerScored player'  // player''.score = 3

> If you have only one module in your file, the `module` can be specified at the top of the file.

Reviewer: ```suggestion
If you have only one module in a file, the `module` can be declared at the top of the file (_without an `=` and an indent_)
```

    module Functions  // notice there is no '=' when at the top of a file

    let addToFive num = 5 + num
    let subtractFive num = num - 5

### Namespaces
Namespaces are simply dotted strings that prefix names. They are placed at the top of the file.

> An inner namespace can be specified in the same file.

Reviewer: This suggests that there is some 'inner namespace' concept. Want to convey that the tooling lets you `open` and/or dot into things based on a hierarchy of namespaces

Author: Rewording coming in an update.

    namespace MyNamespace

>     namespace MyNamespace.InnerSpace

Reviewer: ```suggestion
    namespace MyNamespace.SubNamespace
```

>     namespace MyNamespace.InnerSpace

Reviewer: ```suggestion
    namespace MyNamespace.SubNamespace
```

Author: Fixed in a coming update

Reviewer: maybe an InnerModule and/or a mention of the fact that a module inside a module is actually a nested satatic class IRL and as such inherits visibility constraints from its parents might be possible to work in somewhere?

> They can also be part of a module name.

Reviewer: ```suggestion
They can also be specified in a file-level `module` definition (TODO add ref)
```

Author: Fixing in a coming update.

>     module MyNamespace.InnerSpace.MyModule

Reviewer: ```suggestion
    module MyNamespace.SubNameSpace.MyModule
```

### Open and AutoOpen

The `open` keyword can be used with `module`, `namespace`, and `type`.

    open System.Diagnostics  // open namespace
    let stopwatch = Stopwatch.StartNew()
    ---
    module MyModule =
        type DU1 = A | B | C
        type DU2 = D | E | F

    open type MyModule.DU1
    let du1 = A
    let duNotDefined = D  // 'D' not defined
    open MyModule
    let du2 = D
    ---

>     [<AutoOpen>]

Reviewer: need to convey that this is a global thing, that anyone importing the contaiing namespace will have it implicitly opened (and that you might often do an `open` X immediately after declaring it instead to avoid polluting the golbal namespace with implicit things)

Author: Coming update.

    module MyModule =
        type DU = A | B | C

    let du = A

### Accessibility Modifiers

> F# supports `public`, `private` and `internal`.  It can be applied to `module`, `let`, `member`, and `type`.

Reviewer: says something abut that `protected` elephant!

Author: > says something abut that protected elephant!

Go write C#, you OO heathens!  (I kid). I was torn on it, but if you're doing `protected` you're probably past a cheat-sheet. And no reason to offer bad habits. :)

    module private MyModule = ...
    ---
    let private x = ...
    ---
    member private self.x = ...
    ---
    type private Person = ...

### Recursive Reference

> F#'s dependency resolution is based on file and code order. This requirement encourages developers to think about the design of their programs and dependencies upfront,

Reviewer: ```suggestion
F#'s type inference and name resolution runs in file and line order; by default, any forward references are considered as errors. This default provides a single benefit, which can be hard to apprciate the value of initially: you never need to look beyond the current file, as all items something can depend on must have been declared somewhere before its use. In general this also nudges toward more careful design and organisation of codebases,
```

which results in cleaner, maintainable code, but in rare cases you may need to loosen those rules.

> To do this we have `rec` for `module` and `namespace`s; and `and` for `type`s and `let` functions.

Reviewer: this reads well, but maybe tail end of last sentence needs to be incorporated; perhaps as a separated paragraph?

Author: I was planning on links, but looks like the type...and construct isn't in the cheat sheet yet.

Author: I created a couple more Issues for these things.

    module rec MyNamespace.MonkeyDomain

    exception DoNotSqueezeBananaException of Banana // `Banana` has not been defined yet, and would fail without `rec`

    type Banana = { Type: string; IsRipe: bool }

>         member self.Squeeze() = raise (DoNotSqueezeBananaException self)

Reviewer: better to bring `_` vs `this` vs `self` into play elsewhere
```suggestion
        member _.Squeeze() = raise (DoNotSqueezeBananaException self)
```

Author: We need the `self` to add to `DoNotSqueezeBananaException`. However, after looking at it again, I realize *I* have a bug because I need a `with` declaration before `member`. :-D

Reviewer: ah..., missed that
but... better to format without a `with` by doing
```
type X =
    { .... }
    member self_.Squeeze()
```
(F# Style guide talks about this a bit IIRC)

    and

> See [Namespaces (MS Learn)](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/namespaces) and [Modules (MS Learn)](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/modules) to learn more.

Reviewer: Yes, it's definitely a strange meta-task, second guessing what a beginner would a) need and b) have a chance of inferring from an example while in 'cheat mode'

        | DivisibleBy 3 & DivisibleBy 5 -> "FizzBuzz"
        | DivisibleBy 3 -> "Fizz"
        | DivisibleBy 5 -> "Buzz"
        | i -> string i

*Partial active patterns* share the syntax of parameterized patterns but their active recognizers accept only one argument.

<a name="CodeOrganization"></a>Code Organization
---------

<a name="CodeOrganization_Modules"></a>
### Modules
Modules are key building blocks for grouping related code; they can contain `types`, `let` bindings, or (nested) sub `module`s.
Identifiers within modules can be referenced using dot notation, or you can bring them into scope via the `open` keyword. Illustrative-only example:

    module Game =
        let mutable basePoints = 1
        type Player = { id: int; score: int }
        let playerScored player = { player with score = player.score + basePoints }

    let player: Game.Player = { id = 1; score = 0 }
    let player' = Game.playerScored player  // score = 1

    open Game
    basePoints <- 2
    let player'' = playerScored player'  // player''.score = 3

If you have only one module in a file, the `module` name can be declared at the top of the file, with all further declarations

> being module elements (and non indentation required)

Reviewer: elements is jarring but dont have a better term
"and non indentation required" -> and no indentation required

Author: This was actually your change. :) https://github.com/fsprojects/fsharp-cheatsheet/pull/25#discussion_r1382686659

Reviewer: LOL. This is why I detest my writing!

And good luck with building grammarly in excel :D

Author: A small part of taking this on was to try and improve my own writing. Mixed results so far. :o)

    module Functions  // notice there is no '=' when at the top of a file

    let addToFive num = 5 + num
    let subtractFive num = num - 5

### Namespaces

> Namespaces are simply dotted names that prefix other program elements to allow for further hierarchical organization.

Reviewer: i think the program elements are types and modules, what else?
also maybe prefix -> scope?

Author: Good point, I'll specify as there are only two.

Explaining this simply is proving tricky. Will try again.

> All `type` and `module` elements that follow a `namespace` declaration will require an [`open`](#CodeOrganization_OpenAndAutoOpen) or to be dotted-into to access.

Reviewer: that follow -> within ?

Author: "following in the file" ?

Reviewer: demarcated by :D

I wanted to allude to a `namespace` being a grouping construct that things logically within the scope of

And pave the way for saying "and if you use `namespace` again, then it's an entirely separated new scope with nothing carried over"

I guess module and namespace have that aspect in common

> If a new `namespace` is specified, all elements following will be part of the new namespace.

Reviewer: Any subsequent `namespace` directive in the same file switches to a fresh scope, without carrying forward any `open`s, nesting etc.

    namespace MyNamespace

    namespace MyNamespace.SubNamespace

> They can also be specified in a file-level [`module`](#CodeOrganization_Modules) definition, but no further `namespace` declarations may follow.

Reviewer: A top level module declaration can bear its namespace as a dotted prefix:

Author: I'm not fond of "bear"; but me, the thesaurus, and ChatGPT can't come up with something better.

Author: Maybe "manifest". I'll give you final rule on the PR

    module MyNamespace.SubNamespace.Functions

<a name="CodeOrganization_OpenAndAutoOpen"></a>
### Open and AutoOpen

The `open` keyword can be used with `module`, `namespace`, and `type`.

    open System.Diagnostics  // open namespace
    let stopwatch = Stopwatch.StartNew()
    ---
    module MyModule =
        type DU1 = A | B | C
        type DU2 = D | E | F

    open type MyModule.DU1
    let du1 = A
    let duNotDefined = D  // 'D' not defined
    open MyModule
    let du2 = D

Available to `module` elements, is the `AutoOpen` attribute. This alleviates the need for an `open`; *however* this should be used cautiously,

> as all following declarations will be immediately brought into the global namespace and cause conflicts.

Reviewer: 'following' is a bit ambiguous; maybe all elements/contents
'the global namespace' -> 'the scope whenever the parent is `open`ed, which can cause ambiguities'
(basically its not a conflict, which would be clear and unambiguous - its shadowing, and things become messy as the winning name is defined by the open order)

Perhaps this section should nod to the fact that every open can shadow names, but that in general you want to avoid that coming into play except in very select circumstances.

(People have all sorts of things like topological grouping of opens i.e. system stuff, then domain stuff etc but IME having no shadowing and doing one group sorted alphabetically is probably the best overall)

Other things here are that opens should generally specify the full namespace, i.e. no `open System` then `open Globalization`

    [<AutoOpen>]
    module MyModule =
        type DU = A | B | C

    let du = A

### Accessibility Modifiers

> F# supports `public`, `private` (restraining the element to its containing `type` or `module`) and `internal` (restraining the element to its containing assembly).

Reviewer: I think this is well worded but maybe 'restraining' can become 'limiting usage' or 'limiting references' or 'limiting it's visibility [scope]'?

Author: I'll try "limiting access".

> It can be applied to `module`, `let`, `member`, `type`, and `new`.

Reviewer: It->they/visibility modifiers ?
and primary constructors?

|-------------------------------------------------------------------|--------------------------------------------|
| Module                                                            | `module internal MyModule =`               |
| Module .. `let`                                                   | `let private value =`                      |
| Record                                                            | `type internal MyRecord = { id: int }`     |
| Record [ctor](#CodeOrganization_PrivateConstructors)              | `type MyRecord = private { id: int }`      |
| Discriminated Union                                               | `type private MyDiscUni = A \| B`          |
| Discriminated Union [ctor](#CodeOrganization_PrivateConstructors) | `type MyDiscUni = internal A \| B `        |
| Class                                                             | `type internal MyClass() =`                |
| Class [ctor](#CodeOrganization_PrivateConstructors)               | `type MyClass private () =`                |
| Class Additional [ctor](#CodeOrganization_PrivateConstructors)    | `internal new() = MyClass("defaultValue")` |

> | Class .. `let`                                                    | *Always private. Cannot be modified*       |

Reviewer: but see `val` ?

Author: <-- Noob. I haven't come across `val` before, or I saw them a while ago and forgot. Thanks

Reviewer: to be fair, `val` is fairly esoteric, though it's important/useful in some cases as it is initialized immediately and not calculated each time like a property

Author: Added a new `task..val` Issue as well #29

| `type` .. `member`                                                | `member private _.classMember =`           |

<a name="CodeOrganization_PrivateConstructors"></a>
##### Private Constructors

> Limiting `type` constructors (ctor) accessibility is a good way to enforce value integrity.

Reviewer: maybe this should bring in that there are primary constructors and auxillary ones (`new()`) ?
Not sure about ctor but I guess it makes sense to slip it into a cheatsheet

Author: I created an Issue for myself to add `new` to the class section and then create the link back. For now, I'll have to point at MS Learn.

I'll update the text, but I hesitate to add a link because the F# for Fun and Profit blogpost doesn't actually use the keyword `private`. And, interestingly, his post on single case unions says to use naming conventions to indicate a private members, rather than just using `private` (which is mentioned in the comments and updated in his book). I tried a quick web search but didn't get anything I thought was satisfying.

Reviewer: That's fair - yes, there's a lot of stuff that's pretty involved and being judicious in whether to link to somethig general is definitely an important consideration

Author: I found the content that I think ScottW was pointing at; a YouTube of Yaron Minsky. I'll put that for now.

Example of Single-case Discriminated Union with a `private` constructor:

    type UnitQuantity =
        private UnitQuantity of int
        with
        static member private MaxQty = 100
        static member Create (qty:int) : Result<UnitQuantity, string> =
            if qty <= UnitQuantity.MaxQty
            then Ok (UnitQuantity qty)
            else Error $"UnitQuantity cannot be more than {UnitQuantity.MaxQty}"
    
    let uQty = UnitQuantity.Create 50
    match uQty with
    | Ok (UnitQuantity qty) -> printfn $"Good: {qty}"
    | Error errStr -> printfn $"Bad: {errStr}"

Example of a class with a `private` constructor:

    type MyClass private (count:int) =
        member this.Count = count

>         static member CreateInstance (cnt: int) : MyClass option =

Reviewer: if you had an example with a private primary and a public validating new, then this could be a sister module 
```
module MyClass =
    let tryCreate cnt =
        if cnt <= 0 then None
        else MyClass(cnt)
```
(there should be a section on types with companion modules being a thing vs lots of static members)        
also the `new` here will raise a warning in IDEs as they are only 'required' for things with IDisposable

Author: I just took this out. Seemed redundant and non-idiomatic to show the class side.

            if cnt > 0
            then Some(new MyClass(cnt))
            else None

>     let myClass = MyClass.CreateInstance (5)

Reviewer: space before function invocation args, never for a member invocation's args

### Recursive Reference

F#'s type inference and name resolution runs in file and line order; by default, any forward references are considered as errors.
This default provides a single benefit, which can be hard to appreciate initially: you never need to look beyond the current file for a dependency.
In general this also nudges toward more careful design and organisation of codebases,

> which results in cleaner, maintainable code. However, in rare cases you may need to loosen those rules.

Reviewer: to do some forward references via `and`/`rec`, or, (in extreme circumstances loosen the rules entirely).

Author: I got rid of the "you" reference, but thought this was a bit redundant.

> To do this we have `rec` for `module` and `namespace`s; and `and` for `type` and [`let`](#Functions_RecursiveFunctions) functions.

Reviewer: first, show the main thing to do (use `type` ... `and`)

For me, `rec` shol;d not be flagged to beginners without good explanation as people will be yearning for a "C# mode" and end up frustrated

Author: Added type..and example. Still have to get to the Issue #26; but maybe it is just a link down to this section...

    module rec MyNamespace.MonkeyDomain

>     exception DoNotSqueezeBananaException of Banana // `Banana` has not been defined yet, and would fail without `rec`

Reviewer: this is a valid example, but very esoteric

I'd say most common usage of `rec` is people saying "well for this file, I just want to list everything alphabetically, regardless of the fact that it's normally good to put low level stuff first and then have the hihger level stuff at the bottom, with some exceptions where you use `and` to put some irrelevant dependencies underneath"

So while this does illustrate what it does and how to use it, I dont think it conveys what it's for really, so am pretty conflicted.

Author: I stole this from [MS Learn](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/modules#recursive-modules). I kind of hate it too. The long names get in the way of what's being demonstrated as well.

Author: I took a walk and realized I'm conflicted because I hate exceptions. :)

Reviewer: whereas I think it should have a link to https://eiriktsarpalis.wordpress.com/2017/02/19/youre-better-off-using-exceptions/ and https://fsharpforfunandprofit.com/posts/against-railway-oriented-programming/ :P

| Discriminated Union                                               | `type private MyDiscUni = A \| B`          |
| Discriminated Union [ctor](#CodeOrganization_PrivateConstructors) | `type MyDiscUni = internal A \| B `        |
| Class                                                             | `type internal MyClass() =`                |
| Class [ctor](#CodeOrganization_PrivateConstructors)               | `type MyClass private () =`                |
| Class Additional [ctor](#CodeOrganization_PrivateConstructors)    | `internal new() = MyClass("defaultValue")` |
| Class .. `let`                                                    | *Always private. Cannot be modified*       |
| `type` .. `member`                                                | `member private _.classMember =`           |

<a name="CodeOrganization_PrivateConstructors"></a>
##### Private Constructors

Limiting `type` constructors (ctor) accessibility is a good way to enforce value integrity.

Example of Single-case Discriminated Union with a `private` constructor:

    type UnitQuantity = private UnitQuantity of int
        with

>         static member private maxQty = 100

Author: Going CamelCase.

I tried the indentation style you suggested, but the compiler is complaining unless I (a) indent all the static members a second time; or, (b) add a `with`. I couldn't see anything in the style guide on it either.

>         static member private maxQty = 100

Author: Thanks for all the help on this.
